### PR TITLE
[!feat] Refactor CoinGecko

### PR DESCRIPTION
**Pull request by warittornc**
## Implementation details
- Do 1
- Do 2

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the GitHub PR explorer)

### CODE REVIEW

File: bothan-coingecko/src/api/builder.rs [hunk 1]
```diff
@@ -32,32 +32,44 @@ pub struct CoinGeckoRestAPIBuilder {
 }
 
 impl CoinGeckoRestAPIBuilder {
+    pub fn new<T: Into<String>>(url: Option<T>, api_key: Option<T>, user_agent: T) -> Self {
```
Comment (sorawit) [inline bothan-coingecko/src/api/builder.rs:35]: Should use 3 different types.

File: bothan-coingecko/src/worker.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+use std::sync::Arc;
+
+use bothan_core::store::Store;
+use bothan_core::worker::{AssetStatus, AssetWorker, Error};
+
+use crate::api::CoinGeckoRestAPI;
+
+// use error::Error;
```
Comment (sorawit) [inline bothan-coingecko/src/worker.rs:8]: Remove unused comment

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+use bothan_core::store::Store;
+use bothan_core::types::AssetInfo;
+
+use crate::api::rest::CoinGeckoRestAPI;
+use crate::api::types::{Market, Order};
+use crate::worker::error::ParseError;
+use crate::worker::CoinGeckoWorker;
+
+pub(crate) async fn start_asset_worker(
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:18]: does this function need to be async?

File: bothan-coingecko/src/api/error.rs [hunk 4]
```diff
@@ -1,53 +1,23 @@
...
-
-    /// Failed to parse the response.
-    #[error("failed to parse")]
-    Parse,
-
-    /// An error occurred with the `reqwest` client.
-    #[error("reqwest error: {0}")]
-    Reqwest(String),
-}
+    #[error("received status: {0}")]
```
Comment (sorawit) [inline bothan-coingecko/src/api/error.rs:18]: Please make the error message more descriptive

File: bothan-coingecko/src/api/rest.rs [hunk 5]
```diff
@@ -21,57 +20,54 @@ impl CoinGeckoRestAPI {
...
     /// Retrieves market data for the specified coins from the CoinGecko API.
-    pub async fn get_coins_market(
+    pub async fn get_coins_market<T: AsRef<str>>(
         &self,
-        ids: &[&str],
-        page_size: usize,
-        page: usize,
-    ) -> Result<Vec<Option<Market>>, RestAPIError> {
+        ids: &[T],
+        order: Option<Order>,
```
Comment (sorawit) [inline bothan-coingecko/src/api/rest.rs:31]: I think the caller should always be providing these options. Having default value means the caller may not get prices of all the supplied coins.

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                update_all_asset_info(&store, &api, ids, page_size, page_query_delay),
+            )
+            .await;
+
+            if result.is_err() {
+                warn!("updating interval exceeded timeout")
+            }
+
+            update_interval.tick().await;
+        }
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:43]: Debug log when this function exits.

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    mut update_interval: Interval,
+    page_size: usize,
+    page_query_delay: Duration,
+) {
+    tokio::spawn(async move {
+        while let Some(worker) = weak_worker.upgrade() {
+            info!("updating asset info");
+
+            let ids = worker.store.get_query_ids().await;
+            let store = worker.store.clone();
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:29]: Is clone here necessary? I'm not sure

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bothan-coingecko/src/api/rest.rs (hunk 5 above)
Comment (warittornc) [reply bothan-coingecko/src/api/rest.rs:31]: I've changed it to omit the parameter if None is given to use the endpoints default value as the endpoint doesn't require them.

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    page_query_delay: Duration,
+) {
+    tokio::spawn(async move {
+        while let Some(worker) = weak_worker.upgrade() {
+            info!("updating asset info");
+
+            let ids = worker.store.get_query_ids().await;
+
+            let result = timeout(
+                page_query_delay,
```
Comment (RogerKSI) [inline bothan-coingecko/src/worker/asset_worker.rs:31]: timeout should be at least page_query_delay * page?

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        info!("worker has been dropped, stopping asset worker");
+    });
+}
+
+async fn update_all_asset_info(
+    store: &Store,
+    api: &CoinGeckoRestAPI,
+    ids: Vec<String>,
+    page_size: usize,
+    page_delay: Duration,
```
Comment (RogerKSI) [inline bothan-coingecko/src/worker/asset_worker.rs:52]: I saw the caller function uses `page_query_delay`. please align name.

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: bothan-coingecko/src/api/builder.rs [hunk 10]
```diff
@@ -32,32 +32,50 @@ pub struct CoinGeckoRestAPIBuilder {
 }
 
 impl CoinGeckoRestAPIBuilder {
+    pub fn new<T, U, V>(url: Option<T>, api_key: Option<U>, user_agent: V) -> Self
```
Comment (sorawit) [inline bothan-coingecko/src/api/builder.rs:35]: - I think optional parameters should come after the required argument.
- I think if we want to take `Option` we should not use type param (just use `Option<String>`) otherwise when passing `None` user will need to supply type explicitly which is quite frustrating.

File: bothan-coingecko/src/api/builder.rs [hunk 11]
```diff
@@ -32,32 +32,50 @@ pub struct CoinGeckoRestAPIBuilder {
...
         let mut headers = HeaderMap::new();
-        headers.insert("User-Agent", HeaderValue::from_str(&self.user_agent)?);
+        let agent = match HeaderValue::from_str(&self.user_agent) {
+            Ok(agent) => agent,
+            Err(_) => return Err(BuilderError::InvalidHeaderValue(self.user_agent)),
+        };
+
+        headers.insert("User-Agent", agent);
 
         let url = match self.url {
```
Comment (sorawit) [inline bothan-coingecko/src/api/builder.rs:80]: I think matching both url and api key at the same time as tuple is simpler?

```
| (Some(url), _) =>
| (None, Some(_)) =>
| (None, None) =>
```

File: bothan-coingecko/src/api/builder.rs [hunk 12]
```diff
@@ -69,12 +87,18 @@ impl CoinGeckoRestAPIBuilder {
...
         if let Some(key) = &self.api_key {
-            let mut val = HeaderValue::from_str(key)?;
-            val.set_sensitive(true);
-            headers.insert("x-cg-pro-api-key", val);
+            let mut api_key = match HeaderValue::from_str(key) {
+                Ok(key) => key,
+                Err(_) => return Err(BuilderError::InvalidHeaderValue(key.clone())),
+            };
+            api_key.set_sensitive(true);
+            headers.insert("x-cg-pro-api-key", api_key);
```
Comment (sorawit) [inline bothan-coingecko/src/api/builder.rs:95]: Make this string a constant somewhere

File: bothan-coingecko/src/api/builder.rs [hunk 13]
```diff
@@ -69,12 +87,18 @@ impl CoinGeckoRestAPIBuilder {
...
         }
 
-        let client = ClientBuilder::new().default_headers(headers).build()?;
+        let client = match ClientBuilder::new().default_headers(headers).build() {
+            Ok(client) => client,
```
Comment (sorawit) [inline bothan-coingecko/src/api/builder.rs:99]: I prefer `map_err` with `?` syntax but up to you

File: bothan-coingecko/src/api/types.rs [hunk 14]
```diff
@@ -1,14 +1,37 @@
...
+    MarketCapDesc,
+    VolumeAsc,
+    VolumeDesc,
+    IdAsc,
+    IdDesc,
+}
+
+impl Display for Order {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let str = match self {
```
Comment (sorawit) [inline bothan-coingecko/src/api/types.rs:23]: str can just be `&'static str` and we don't need string allocation?

File: bothan-coingecko/src/worker.rs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+mod asset_worker;
+pub mod builder;
+pub mod error;
+pub mod opts;
+pub mod types;
+
+/// A worker that fetches and stores the asset information from CoinGecko's API.
+pub struct CoinGeckoWorker {
+    api: Arc<CoinGeckoRestAPI>,
```
Comment (sorawit) [inline bothan-coingecko/src/worker.rs:16]: Do we need Arc here? Can we not use `CoinGeckoRestAPI` directly?

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+            .await;
+
+            if result.is_err() {
+                warn!("updating interval exceeded timeout")
+            }
+
+            update_interval.tick().await;
+        }
+
+        info!("worker has been dropped, stopping asset worker");
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:43]: should we do `debug` so it's consistent with Binance's asset worker? also let's make it's the same message.

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+async fn update_all_asset_info(
+    store: &Store,
+    api: &CoinGeckoRestAPI,
+    ids: Vec<String>,
+    page_size: usize,
+    page_query_delay: Duration,
+) {
+    let mut interval = interval(page_query_delay);
+    let queue = Vec::from_iter(ids.chunks(page_size));
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:55]: do we need to create a vector? can we iterate the chunk?

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    for ids in queue {
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:56]: Let's not use the same variable name

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+    api: &CoinGeckoRestAPI,
+    ids: &[String],
+    page_size: usize,
+) {
+    match api
+        .get_coins_market(ids, Some(Order::IdDesc), Some(page_size), Some(1))
+        .await
+    {
+        Ok(markets) => {
+            let to_set = markets
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:73]: Should we sanity check that the number of markets we get is >= the id size? better to know if sth unexpected happen fast

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+
+pub(crate) fn start_asset_worker(
+    weak_worker: Weak<CoinGeckoWorker>,
+    mut update_interval: Interval,
+    page_size: usize,
+    page_query_delay: Duration,
+) {
+    tokio::spawn(async move {
+        while let Some(worker) = weak_worker.upgrade() {
+            info!("updating asset info");
```
Comment (sorawit) [inline bothan-coingecko/src/worker/asset_worker.rs:26]: will these logs prefixed with coingecko somewhere? do we need to creating tracing span? same for binance. can be a separate PR

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bothan-coingecko/src/worker/asset_worker.rs [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+use bothan_core::types::AssetInfo;
+
+use crate::api::rest::CoinGeckoRestAPI;
+use crate::api::types::{Market, Order};
+use crate::worker::error::ParseError;
+use crate::worker::CoinGeckoWorker;
+
+pub(crate) fn start_asset_worker(
+    weak_worker: Weak<CoinGeckoWorker>,
+    mut update_interval: Interval,
```
Comment (taobun) [inline bothan-coingecko/src/worker/asset_worker.rs:20]: Umm, can we receive `Duration` from caller and create `Interval` in this function instead?

File: bothan-coingecko/src/api/rest.rs [hunk 22]
```diff
@@ -21,57 +20,56 @@ impl CoinGeckoRestAPI {
...
+        page_size: Option<usize>,
+        page: Option<usize>,
+    ) -> Result<Vec<Market>, RestAPIError> {
         let url = format!("{}coins/markets", self.url);
```
Comment (taobun) [inline bothan-coingecko/src/api/rest.rs:35]: What 's the different between these 2 endpoints?
1. https://docs.coingecko.com/reference/simple-price
2. https://docs.coingecko.com/reference/coins-markets

Comment (warittornc) [reply bothan-coingecko/src/api/rest.rs:35]: coins-market is more verbose, but since simple-price is newer and doesn't require pagination, will switch to that instead

File: bothan-coingecko/src/api/builder.rs [hunk 23]
```diff
@@ -26,55 +26,76 @@ use crate::api::CoinGeckoRestAPI;
 /// }
 /// ```
 pub struct CoinGeckoRestAPIBuilder {
+    user_agent: String,
     url: Option<String>,
```
[the remark this answers is not in the document]
Comment (warittornc) [reply bothan-coingecko/src/api/builder.rs:30]: this is intended to distinguish if the user has overidden the value

File: bothan-coingecko/examples/coingecko_basic.rs [hunk 24]
```diff
@@ -2,20 +2,21 @@ use std::time::Duration;
...
+use bothan_core::worker::AssetWorker;
 
 #[tokio::main]
 async fn main() {
     init();
-    let service_result = CoinGeckoServiceBuilder::default()
-        .with_update_supported_assets_interval(Duration::from_secs(600))
-        .build()
-        .await;
+    let worker_result = CoinGeckoWorkerBuilder::default().build().await;
```
Comment (colmazia) [inline bothan-coingecko/examples/coingecko_basic.rs:11]: Why name it worker_result na, it's a working worker with no query_ids right?

Comment (colmazia) [reply bothan-coingecko/examples/coingecko_basic.rs:11]: Perhaps default_worker?

File: bothan-coingecko/src/api/builder.rs [hunk 25]
```diff
@@ -26,55 +26,76 @@ use crate::api::CoinGeckoRestAPI;
...
-                None => DEFAULT_URL.to_string(),
-                Some(_) => DEFAULT_PRO_URL.to_string(),
-            },
+        let url = match (&self.url, &self.api_key) {
+            (Some(url), _) => url,
+            (None, Some(_)) => DEFAULT_PRO_URL,
+            (None, None) => DEFAULT_URL,
         };
-        let parsed_url = Url::parse(&url)?;
+        let parsed_url = Url::parse(url)?;
```
Comment (colmazia) [inline bothan-coingecko/src/api/builder.rs:84]: Why not map this err to BuildError like others na?

File: bothan-coingecko/src/api/builder.rs [hunk 26]
```diff
@@ -84,7 +105,7 @@ impl Default for CoinGeckoRestAPIBuilder {
     /// Creates a default `CoinGeckoRestAPIBuilder` instance with default values.
     fn default() -> Self {
         CoinGeckoRestAPIBuilder {
-            url: None,
+            url: Some(DEFAULT_URL.into()),
```
Comment (colmazia) [inline bothan-coingecko/src/api/builder.rs:108]: So we expect the user with api_key to change the url themself right?

File: bothan-coingecko/src/api/builder.rs (hunk 25 above)
Comment (colmazia) [reply bothan-coingecko/src/api/builder.rs:84]: Is it auto parsed to BuildError? Since I see there is InvalidURL in BuildError

File: bothan-coingecko/src/api/error.rs [hunk 27]
```diff
@@ -1,53 +1,23 @@
...
-}
+    #[error("received non-2xx http status: {0}")]
+    UnsuccessfulResponse(reqwest::StatusCode),
 
-impl From<reqwest::Error> for RestAPIError {
-    fn from(e: reqwest::Error) -> Self {
-        RestAPIError::Reqwest(e.to_string())
-    }
+    #[error("failed to parse with error: {0}")]
+    ParseFailed(String),
```
Comment (colmazia) [inline bothan-coingecko/src/api/error.rs:22]: You mean parse the response? Maybe add more detail.

File: bothan-coingecko/src/worker.rs [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+        self.store.get_assets(ids).await
+    }
+
+    /// Adds the specified cryptocurrency IDs to the query set and subscribes to their updates.
+    async fn add_query_ids<K: Into<String> + Send + Sync>(&self, ids: Vec<K>) -> Result<(), Error> {
+        self.store.add_query_ids(ids).await;
+        Ok(())
+    }
+
+    /// Removes the specified cryptocurrency IDs to the query set and subscribes to their updates.
```
Comment (colmazia) [inline bothan-coingecko/src/worker.rs:40]: last part of the comment is incorrect, we don't subscribe to removed query ids.

File: bothan-coingecko/src/api/builder.rs (hunk 26 above)
Comment (warittornc) [reply bothan-coingecko/src/api/builder.rs:108]: default is without an api key, if the user supplies an api key it will auto-change the url to the pro endpoint if it hasn't been set by the user. See line 79-84 for reference

File: bothan-coingecko/src/api/error.rs (hunk 27 above)
Comment (warittornc) [reply bothan-coingecko/src/api/error.rs:22]: this error is unique to function next(), being more verbose here is not necessary in my opinion

File: bothan-coingecko/src/api/builder.rs (hunk 26 above)
Comment (colmazia) [reply bothan-coingecko/src/api/builder.rs:108]: Yep, but doesn't put value of some here contradict that purpose, because if user call default function and then put an api key, it will enter case of line 80 which is not a pro url.

Comment (warittornc) [reply bothan-coingecko/src/api/builder.rs:108]: oh right, it should be None, fixed

**Review by colmazia — APPROVED** [review]

**Review by RogerKSI — APPROVED** [review]

**Review by taobun — APPROVED** [review]
